libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp

```diff
@@ -1044,3 +1048,25 @@ float NavEKF3_core::MagDeclination(void) const
     }
     return _ahrs->get_compass()->get_declination();
 }
+
+/*
+  update Z gyro bias. This is only used for
+  MagCal::EXTERNAL_YAW_FALLBACK, when not flying and when we have not
+  yet done a yaw alignment. It prevents a buildup of a large Z gyro
+  bias while waiting for GPS yaw
+ */
+void NavEKF3_core::updateZGyroBias(void)
+{
+    const float zgyro_learn_limit = radians(3);
+    const AP_InertialSensor &ins = AP::ins();
+    float &zbias = stateStruct.gyro_bias.z;
+    float z_gyro = ins.get_gyro(gyro_index_active).z - (zbias / dtEkfAvg);
+    if (fabsf(z_gyro) > zgyro_learn_limit) {
```

Comment:
This simple threshold check would be easily fooled if starting from the deck of a boat where slow yaw drift is common and also checking Z is not applicable for tailsitters. 

- Rather than updating Z we should be updating the component aligned within TBD of the gravity vector .

- Gyro rate data from the orthogonal axes should then also be used in the motion check which will reduce the likelihood of false triggering when taking off from a boat deck.

Reply:
I've fixed it to use the components for earth frame yaw, I've bench tested it for a variety of attitudes on the bench, including roll 90 and pitch 90, and checked that it zeros an artificially induced gyro bias about the relevant axis.

Reply:
This cam still run into problems with slow yaw drift, eg boats at moorings. There is a better method, and that is to use the same motion check that the AHRS uses when doing the startup gyro cal tht is more sensitive to movements. A class variable containing the last yaw angle when movement is detected is updated each frame. When no movement is dtected, that yaw angle is then used a yaw observation by the EKF. This ensures that bias states and  covariances  are updated correctly regardless of orientation. Will call to discuss.